packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -42,105 +141,293 @@ export function createElement(original: parse5.Element): IRElement {
+export function isBaseElement(node: BaseNode): node is Element {
+    return isElement(node) || isComponent(node) || isSlot(node);
 }
 
-export function isSlot(element: IRElement) {
-    return element.tag === 'slot';
+export function isTextNode(node: BaseNode): node is Text {
+    return node.type === LWCNodeType.Text;
 }
 
-export function isIRExpressionAttribute(
-    attribute: IRAttribute
-): attribute is IRExpressionAttribute {
-    return attribute.type === IRAttributeType.Expression;
+export function isCommentNode(node: BaseNode): node is Comment {
+    return node.type === LWCNodeType.Comment;
 }
 
-export function isIRStringAttribute(attribute: IRAttribute): attribute is IRStringAttribute {
-    return attribute.type === IRAttributeType.String;
+export function isTemplate(node: ParentNode): boolean {
+    return node.name === 'template';
 }
 
-export function isIRBooleanAttribute(attribute: IRAttribute): attribute is IRBooleanAttribute {
-    return attribute.type === IRAttributeType.Boolean;
+export function isExpression(node: Expression | Literal): node is Expression {
+    return node.type === LWCNodeType.Identifier || node.type === LWCNodeType.MemberExpression;
 }
 
-export function isComponentProp(
-    identifier: TemplateIdentifier,
-    root: IRNode,
-    parentStack: IRNode[]
-): boolean {
-    const { name } = identifier;
-    let current: IRNode | undefined = root;
+export function isStringLiteral(node: Expression | Literal): node is Literal<string> {
+    return node.type === LWCNodeType.Literal && typeof node.value === 'string';
+}
 
-    // Walking up the AST and checking for each node to find if the identifer name is identical to
-    // an iteration variable.
-    for (let i = parentStack.length; i >= 0; i--) {
-        if (isElement(current)) {
-            const { forEach, forOf } = current;
+export function isBooleanLiteral(node: Expression | Literal): node is Literal<boolean> {
+    return node.type === LWCNodeType.Literal && typeof node.value === 'boolean';
+}
 
-            if (
-                forEach?.item.name === name ||
-                forEach?.index?.name === name ||
-                forOf?.iterator.name === name
-            ) {
-                return false;
-            }
-        }
+export function isForOf(node: BaseNode): node is ForOf {
+    return node.type === LWCNodeType.ForOf;
+}
 
-        current = parentStack[i - 1];
-    }
+export function isForEach(node: BaseNode): node is ForEach {
+    return node.type === LWCNodeType.ForEach;
+}
+
+export function isForBlock(node: BaseNode): node is ForBlock {
+    return isForOf(node) || isForEach(node);
+}
+
+export function isIfBlock(node: BaseNode): node is IfBlock {
+    return node.type === LWCNodeType.IfBlock;
+}
+
+export function isParentNode(node: BaseNode): node is ParentNode {
+    return isBaseElement(node) || isRoot(node) || isForBlock(node) || isIfBlock(node);
+}
+
+export function isProperty(node: BaseNode): boolean {
```

Comment:
```suggestion
export function isProperty(node: BaseNode): node is Property {
```

Reply:
Question about using these type guards.  Is it good practice to use type guards such as `isProperty` to narrow down a type?  I use these type guards to narrow down from `ParentNode` quite often but I'm not sure if it should be avoided when possible.

Reply:
> Is it good practice to use type guards such as `isProperty` to narrow down a type?

Yes type guards are the recommend way to narrow down types. If you give me a list of type guards you are unsure about I can give you a detailed review for each of them.